CB-15139 Modifications to make Periscope differentiate between the sc…

aakashshankar opened the pull request:
…aling mechanism.

### Overview
- Schema change to accommodate stopStartScaling enabled flag for the periscope cluster.
- Changes to ScalingRequest and ClusterStatusSyncHandler to call relevant API based on scaling mechanism.
- **Added new required property to `DistroxAutoscaleClusterRequest` for enabling/disabling stopStart scaling**

### Testing
- Unit tests for the modified components.

### TODO:
- [x] Changes to YarnLoadEvaluator to filter out STOPPED or otherwise UNHEALTHY nodes.
- [x] Determine if the `stopStartScalingEnabled` flag must be exposed in the autoscale API, to toggle between normal autoscaling and StopStart scaling.
- [x] Tests for `DistroXAutoScaleClusterV1Endpoint`
- [ ] ~Add entitlement check for StopStart scaling~


See detailed description in the commit message.

cegganesh84 commented:
LGTM. stopstart_enabled DB flag could be named as stop_start_enabled to be consistent with Java coding.

Removing the entitlement is a laborious process in production, so the API to enable it or disable it would be useful. CB-15140

aakashshankar commented:
> LGTM. stopstart_enabled DB flag could be named as stop_start_enabled to be consistent with Java coding.
Removing the entitlement is a laborious process in production, so the API to enable it or disable it would be useful. CB-15140

Updated DB flag to `stop_start_enabled`, and also added API endpoint to toggle stopStart scaling

cegganesh84 reviewed (approved):
LGTM.

sidseth commented on autoscale/src/main/java/com/sequenceiq/periscope/utils/StackResponseUtils.java:
```diff
@@ -33,9 +33,7 @@
-                // TODO CB-14929: the computation of host group counts depends on the scaling mechanism being used.
-                //  Even for STOPPED instances, this may require additional checks on status of the nodes.
-                .filter(instanceMetaData -> instanceMetaData.getInstanceStatus() != InstanceStatus.STOPPED)
+                .filter(instanceMetaData -> InstanceStatus.SERVICES_HEALTHY.equals(instanceMetaData.getInstanceStatus()))
```
This will need additional changes. Not sure if planned in this PR or a separate PR.

Ref: the comment in YarnLoadEvaluator.

The same method is used to find hosts when triggering downscale and upscale, and at this point also for stop-start vs regular scaling. Both need to be thought through independently.

Would be good to map this out given potential states. From the YarnLoadEvaluator comment (assuming SERVICES_RUNNING instead of SERVICES_STARTING) - this leads to unnecessary scale operations.

aakashshankar replied:
Yes, as discussed, will address the different possibilities in StackResponseUtils and YarnLoadEvaluator as part of a separate PR

sidseth commented on autoscale/src/main/resources/schema/app/20211220140443_CB-15139_Support_for_StopStart_scaling_in_periscope.sql:
```diff
@@ -0,0 +1,12 @@
+-- Migration SQL that makes the change goes here.
+ALTER TABLE "cluster" ADD COLUMN IF NOT EXISTS "stop_start_enabled" boolean;
+ALTER TABLE "cluster" ALTER COLUMN "stop_start_enabled" SET DEFAULT false;
+UPDATE "cluster" SET "stop_start_enabled" = false WHERE "stop_start_enabled" IS NULL;
```
Can we treat NULL values as false? Instead of ALTERING every row in the table?

aakashshankar replied:
Yes, updated to a simpler schema

aakashshankar commented on autoscale-api/src/main/java/com/sequenceiq/periscope/api/model/DistroXAutoscaleClusterResponse.java:
```diff
@@ -98,4 +101,12 @@ public String getStackCrn() {
         this.stackCrn = stackCrn;
     }
+
+    public Boolean getStopStartScalingEnabled() {
```
updated the method name

aakashshankar commented on autoscale-api/src/main/java/com/sequenceiq/periscope/api/endpoint/v1/DistroXAutoScaleClusterV1Endpoint.java:
```diff
@@ -84,4 +85,18 @@ DistroXAutoscaleClusterResponse updateAutoscaleConfigByClusterName(@PathParam("n
     DistroXAutoscaleClusterResponse enableAutoscaleForClusterCrn(@PathParam("crn") String clusterCrn, AutoscaleClusterState autoscaleState);
+
+    @POST
+    @Path("name/{name}/stop_start_scaling")
```
Updated to remove the new API endpoints. Accomodated the stopStart scaling flag within the existing enableAutoScaleForClusterName/enableAutoscaleForClusterCrn endpoints by adding it to the `AutoscaleClusterState.java`:
https://github.com/hortonworks/cloudbreak/blob/8686e26598fb87e9bc66b727fee89a8f9f1e6ebb/autoscale-api/src/main/java/com/sequenceiq/periscope/api/model/AutoscaleClusterState.java#L14-L17.

Also made corresponding modifications to update the autoscaling state for enable disable as well as enableStopStart/disableStopStart within a transaction.

aakashshankar commented on autoscale/src/main/java/com/sequenceiq/periscope/monitor/handler/ClusterStatusSyncHandler.java:
```diff
@@ -42,17 +42,18 @@ public void onApplicationEvent(ClusterStatusSyncEvent event) {
-//                && Optional.ofNullable(statusResponse.getClusterStatus()).map(Status::isAvailable).orElse(false);
+        boolean clusterAvailable;
+        boolean clusterNodesUnhealthy = false;
+        if (cluster.isStopStartScalingEnabled()) {
```
Since this would require changes coming in from #12026 , will examine the possibilities more carefully, and create a separate PR as part of [CB-15141](https://jira.cloudera.com/browse/CB-15141)

aakashshankar commented:
Updated TODO to add entitlement check, since the entitlement changes are present as part of #12026 , will update this PR once it has been merged

sidseth commented on autoscale/src/main/java/com/sequenceiq/periscope/controller/DistroXAutoScaleClusterV1Controller.java:
```diff
@@ -138,6 +138,18 @@ private DistroXAutoscaleClusterResponse createClusterJsonResponse(Cluster cluste
     }
 
+    private DistroXAutoscaleClusterResponse updateClusterAutoScaleState(Cluster cluster, AutoscaleClusterState autoscaleState) {
+        // TODO CB-15139 Use alertValidator to check for StopStart scaling entitlement
```
Could you please file this as a new jira so that it doesn't get lost as a generic TODO

aakashshankar replied:
Yes, filed as a separate JIRA and updated TODO comment: [CB-15673](https://jira.cloudera.com/browse/CB-15673)

sidseth commented on autoscale-api/src/main/java/com/sequenceiq/periscope/api/model/AutoscaleClusterState.java:
```diff
@@ -11,6 +12,9 @@
     @ApiModelProperty(ClusterAutoscaleState.ENABLE_AUTOSCALING)
     private boolean enableAutoscaling;
 
+    @ApiModelProperty(ClusterStopStartScalingState.ENABLE_STOP_START_SCALING)
```
How are defaults being handled here? Specifically, if enableAutoscaleForClusterName (DistroXAutoScaleClusterV1Endpoint) gets invoked with either enableAutoScale set or useStopStartScaling set (not both) - will this end up resetting the value of the other parameter?

The API should allow setting these values independently. If I'm not mistaken, calling it with a single parameter will end up using the default value for the other one - which essentially resets existing state.

aakashshankar replied:
Yes, upon providing only one parameter, for ex: `enableAutoscaling` as true, the other parameter, `useStopStartMechanism` is set to its default value: false, which ends up unintentionally changing the existing state. Updated the implementation so that when only one of these params is provided, the other one isn't altered, i.e, the values can be toggled independently.

sidseth commented on autoscale-api/src/main/java/com/sequenceiq/periscope/api/model/AutoscaleClusterState.java:
```diff
@@ -19,6 +23,14 @@ public void setEnableAutoscaling(boolean enableAutoscaling) {
         this.enableAutoscaling = enableAutoscaling;
     }
 
+    public boolean isUseStopStartMechanism() {
```
Nit: Rename to getUseStopStartMechanism (is* made more sense on the name in the last patch)

aakashshankar replied:
Updated method name

sidseth reviewed (commented):
Some questions on handling API calls when only a single parameter is set. Other than that, looks good.
